File: rust/calc.rs
use std::env;
use std::io::{self, Write};
use std::fs::OpenOptions;
use std::path::PathBuf;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

thread_local! {
    static CONTINUE_FLAG: Arc<AtomicBool> = Arc::new(AtomicBool::new(false));
}

enum Mode {
    Argument,
    InlineSingle,
    InlineMultiple
}

#[allow(dead_code)]
enum Modifier {
    S,
    M,
    A
}

fn apply_operation(a: f64, b: f64, op: char) -> Result<f64, String> {
    match op {
        '+' => Ok(a + b),
        '-' => Ok(a - b),
        '*' => Ok(a * b),
        '/' => {
            if b == 0.0 {
                Err("Error: Division by zero!".to_string())
            } else {
                Ok(a / b)
            }
        }
        '^' => Ok(a.powf(b)),
        _ => Err(format!("Error: Unknown operator '{}'!", op)),
    }
}

fn apply_function(fn_name: &str, x: f64) -> Result<f64, String> {
    match fn_name {
        "sqrt" => Ok(x.sqrt()),
        "log" | "ln" => {
            if x <= 0.0 {
                Err("Error: log domain error!".to_string())
            } else {
                Ok(x.ln())
            }
        }
        "sin" => Ok(x.sin()),
        "cos" => Ok(x.cos()),
        "tan" => Ok(x.tan()),
        _ => Err(format!("Error: Unknown function '{}'!", fn_name)),
    }
}

fn get_precedence(op: char) -> Option<i32> {
    match op {
        '+' | '-' => Some(1),
        '*' | '/' => Some(2),
        '^' => Some(3),
        _ => None,
    }
}

fn check_parentheses(s: &str) -> bool {
    let mut balance = 0;
    for ch in s.chars() {
        match ch {
            '(' => balance += 1,
            ')' => {
                balance -= 1;
                if balance < 0 {
                    return false;
                }
            }
            _ => {}
        }
    }
    balance == 0
}

fn tokenize(eq: &str) -> Vec<String> {
    let mut tokens = Vec::new();
    let mut cur = String::new();
    let chars: Vec<char> = eq.chars().collect();
    let mut i = 0;

    let flush = |tokens: &mut Vec<String>, cur: &mut String| {
        if !cur.is_empty() {
            tokens.push(cur.clone());
            cur.clear();
        }
    };

    while i < chars.len() {
        let ch = chars[i];

        if ch.is_whitespace() {
            i += 1;
            continue;
        }

        // Identifier (function name)
        if ch.is_alphabetic() {
            flush(&mut tokens, &mut cur);
            let mut name = String::new();
            while i < chars.len() && chars[i].is_alphabetic() {
                name.push(chars[i]);
                i += 1;
            }
            tokens.push(name);
            continue;
        }

        // Part of a number (digit, dot, or leading minus)
        if ch.is_numeric() || ch == '.' || (ch == '-' && (
            i == 0 ||
            chars[i - 1] == '(' ||
            get_precedence(chars[i - 1]).is_some()
        )) {
            cur.push(ch);
            i += 1;
        } else {
            flush(&mut tokens, &mut cur);
            tokens.push(ch.to_string());
            i += 1;
        }
    }
    flush(&mut tokens, &mut cur);
    tokens
}

fn evaluate_expression(tokens: &[String], idx: &mut usize) -> Result<f64, String> {
    let mut nums: Vec<f64> = Vec::new();
    let mut ops: Vec<char> = Vec::new();

    while *idx < tokens.len() && tokens[*idx] != ")" {
        let tok = &tokens[*idx];

        // Function call
        if tok.chars().next().map_or(false, |c| c.is_alphabetic()) {
            let fn_name = tok.clone();
            *idx += 1;
            if *idx >= tokens.len() || tokens[*idx] != "(" {
                return Err(format!("Error: Expected '(' after {}", fn_name));
            }
            *idx += 1;
            let arg = evaluate_expression(tokens, idx)?;
            nums.push(apply_function(&fn_name, arg)?);
            continue;
        }

        // Sub-expression
        if tok == "(" {
            *idx += 1;
            nums.push(evaluate_expression(tokens, idx)?);
            continue;
        }

        // Number literal
        if tok.chars().next().map_or(false, |c| c.is_numeric() || (c == '-' && tok.len() > 1)) {
            nums.push(tok.parse::<f64>().map_err(|_| format!("Error: Invalid number '{}'", tok))?);
            *idx += 1;
            continue;
        }

        // Operator
        if tok.len() == 1 {
            if let Some(op_char) = tok.chars().next() {
                if let Some(p1) = get_precedence(op_char) {
                    while let Some(&top_op) = ops.last() {
                        if let Some(p2) = get_precedence(top_op) {
                            let should_pop = if op_char == '^' {
                                p1 < p2
                            } else {
                                p1 <= p2
                            };
                            if should_pop {
                                ops.pop();
                                let b = nums.pop().ok_or("Error: Invalid expression")?;
                                let a = nums.pop().ok_or("Error: Invalid expression")?;
                                nums.push(apply_operation(a, b, top_op)?);
                            } else {
                                break;
                            }
                        } else {
                            break;
                        }
                    }
                    ops.push(op_char);
                    *idx += 1;
                    continue;
                }
            }
        }

        return Err(format!("Error: Invalid token '{}'", tok));
    }

    // Consume ')'
    if *idx < tokens.len() && tokens[*idx] == ")" {
        *idx += 1;
    }

    // Flush remaining ops
    while let Some(op) = ops.pop() {
        let b = nums.pop().ok_or("Error: Invalid expression")?;
        let a = nums.pop().ok_or("Error: Invalid expression")?;
        nums.push(apply_operation(a, b, op)?);
    }

    if nums.is_empty() {
        return Err("Error: Empty expression!".to_string());
    }
    Ok(nums[0])
}

fn evaluate(tokens: &[String]) -> Result<f64, String> {
    let mut idx = 0;
    evaluate_expression(tokens, &mut idx)
}

fn write_history(expr: &str) {
    if expr.is_empty() {
        return;
    }
    if let Ok(home) = env::var("HOME") {
        let path = PathBuf::from(home).join(".calchistory");
        if let Ok(mut file) = OpenOptions::new().create(true).append(true).open(&path) {
            let _ = writeln!(file, "{}", expr);
        }
    }
}

fn ask_continue() {
    print!("Continue? (Y/n): ");
    let _ = io::stdout().flush();
    let mut input = String::new();
    let _ = io::stdin().read_line(&mut input);
    let response = input.trim();
    if response == "n" || response == "N" {
        CONTINUE_FLAG.with(|flag| flag.store(true, Ordering::Relaxed));
        println!("Exiting...");
    }
}

fn argument_mode(args: &str) {
    write_history(args);
    match check_parentheses(args) {
        false => eprintln!("Error: Mismatched parentheses."),
        true => {
            let tokens = tokenize(args);
            match evaluate(&tokens) {
                Ok(result) => println!("{}", result),
                Err(e) => eprintln!("{}", e),
            }
        }
    }
}

fn inline_single_mode() {
    let mut line = String::new();
    print!("Enter expression: ");
    let _ = io::stdout().flush();
    if io::stdin().read_line(&mut line).is_ok() {
        let line = line.trim();
        write_history(line);
        match check_parentheses(line) {
            false => eprintln!("Error: Mismatched parentheses."),
            true => {
                let tokens = tokenize(line);
                match evaluate(&tokens) {
                    Ok(result) => println!("{}", result),
                    Err(e) => eprintln!("{}", e),
                }
            }
        }
    }
}

fn inline_multiple_mode() {
    loop {
        let should_exit = CONTINUE_FLAG.with(|flag| flag.load(Ordering::Relaxed));
        if should_exit {
            break;
        }
        let mut line = String::new();
        print!("Enter expression: ");
        let _ = io::stdout().flush();
        if io::stdin().read_line(&mut line).is_err() {
            break;
        }
        let line = line.trim();
        write_history(line);
        match check_parentheses(line) {
            false => eprintln!("Error: Mismatched parentheses."),
            true => {
                let tokens = tokenize(line);
                match evaluate(&tokens) {
                    Ok(result) => println!("{}", result),
                    Err(e) => eprintln!("{}", e),
                }
            }
        }
        println!("------------------------");
        ask_continue();
    }
}

fn main() {
    let argv: Vec<String> = env::args().collect();
    let argc = argv.len();

    let first_arg = if argc > 1 {
        argv[1].clone()
    } else {
        String::new()
    };

    let mut mode: Mode = Mode::InlineSingle;

    let _saw_set_var: bool = false;
    let _saw_clear_var: bool = false;
    let mut flag_count: i32 = 0;
    for i in 1..argc {
        let a: String = argv[i].clone();
        if a == "-s" {
            mode = Mode::InlineSingle;
            flag_count += 1;
            continue;
        }
        if a == "-m" {
            mode = Mode::InlineMultiple;
            flag_count += 1;
            continue;
        }
        if a == "-a" {
            mode = Mode::Argument;
            flag_count += 1;
            continue;
        }
    }
    
    if flag_count > 1 {
        eprintln!("Error: Multiple mode flags provided.");
        std::process::exit(1);
    }

    if flag_count == 0 && argc > 1 {
        let mut args = Vec::new();
        for i in 1..argc {
            args.push(argv[i].clone());
        }
        let combined = args.join(" ");
        argument_mode(&combined);
        return;
    }

    match mode {
        Mode::InlineSingle => inline_single_mode(),
        Mode::InlineMultiple => inline_multiple_mode(),
        Mode::Argument => {
            // Build expression from non-flag argv elements (ignore known flags)
            let mut args = Vec::new();
            for i in 1..argc {
                let a = &argv[i];
                if !a.is_empty() && a.starts_with('-') {
                    continue; // skip flags
                }
                args.push(a.clone());
            }
            
            if !args.is_empty() {
                let combined = args.join(" ");
                argument_mode(&combined);
            } else if !first_arg.is_empty() {
                argument_mode(&first_arg);
            } else {
                // No argument provided; fall back to inline single prompt
                inline_single_mode();
            }
        }
    }
}

File: cplusplus/calc.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <cctype>
#include <csignal>
#include <vector>
#include <stack>
#include <cmath>
#include <map>
#include <limits>
#include <stdexcept>
#include <fstream>
#include <cstdlib>

class History {
private:
    std::string path;
    std::ostringstream dataCached;

    std::string getHistoryPath(){
        const char* home = std::getenv("HOME");
        if (!home) return "\0";
        return std::string(home) + "/.calchistory";
    }

public:
    History() : path(this->getHistoryPath()) {}

    // Append expression (only) to ~/.calchistory. If HOME is not available or
    // the file cannot be opened, this is a no-op.
    void write(const std::string& expr) {
        if (expr.empty()) return;
        dataCached << expr << '\n';
    }

    void flush(){
        std::ofstream ofs(path);
        ofs << dataCached.str();
        ofs.close();
    }
};

// Global pointer to History for signal handler access
History* g_history = nullptr;

// Calculator run modes
enum class Mode {
    MODE_ARGUMENT,
    MODE_INTERACTIVE
};

void signalHandler(int sig) {
    if (g_history) {
        g_history->flush();
    }
    exit(sig);
}

// Apply binary operators: +, -, *, /, ^
double applyOperation(double a, double b, char op) {
    switch (op) {
    case '+': return a + b;
    case '-': return a - b;
    case '*': return a * b;
    case '/':
        if (b == 0.0) throw std::runtime_error("Error: Division by zero!");
        return a / b;
    case '^': return std::pow(a, b);
    default: throw std::runtime_error(std::string("Error: Unknown operator '") + op + "'!");
    }
}

// Apply unary functions: sqrt, log, sin, cos, tan
double applyFunction(const std::string& fn, double x) {
    if (fn == "sqrt") return std::sqrt(x);
    if (fn == "log" || fn == "ln") {
        if (x <= 0.0) throw std::runtime_error("Error: log domain error!");
        return std::log(x);
    }
    if (fn == "sin") return std::sin(x);
    if (fn == "cos") return std::cos(x);
    if (fn == "tan") return std::tan(x);
    throw std::runtime_error("Error: Unknown function '" + fn + "'!");
}

// Operator precedence
std::map<char, int> precedence = {
    {'+', 1}, {'-', 1},
    {'*', 2}, {'/', 2},
    {'^', 3}
};

// Check balanced parentheses
bool checkParentheses(const std::string& s) {
    int bal = 0;
    for (char ch : s) {
        if (ch == '(') ++bal;
        else if (ch == ')' && --bal < 0) return false;
    }
    return bal == 0;
}

// Split the input into tokens: numbers, operators, parentheses, names
std::vector<std::string> tokenize(const std::string& eq) {
    std::vector<std::string> tokens;
    std::string cur;
    auto flush = [&]() {
        if (!cur.empty()) {
            tokens.push_back(cur);
            cur.clear();
        }
        };
    for (size_t i = 0; i < eq.size(); ++i) {
        char ch = eq[i];
        if (std::isspace(ch)) continue;

        // Identifier (function name)
        if (std::isalpha(ch)) {
            flush();
            std::string name;
            while (i < eq.size() && std::isalpha(eq[i])) {
                name += eq[i++];
            }
            tokens.push_back(name);
            --i;  // back up one char
        }
        // Part of a number (digit, dot, or leading minus)
        else if (std::isdigit(ch) || ch == '.' ||
            (ch == '-' && (
                i == 0 ||
                eq[i - 1] == '(' ||
                precedence.count(eq[i - 1]) > 0
                ))) {
            cur += ch;
        }
        else {
            flush();
            tokens.push_back(std::string(1, ch));
        }
    }
    flush();
    return tokens;
}

// Evaluate expression with recursion for parentheses and functions
double evaluateExpression(
    std::vector<std::string>::const_iterator& it,
    const std::vector<std::string>::const_iterator& end)
{
    std::stack<double> nums;
    std::stack<char> ops;

    while (it != end && *it != ")") {
        const std::string& tok = *it;

        // Function call
        if (std::isalpha(tok[0])) {
            std::string fn = tok;
            ++it;  // skip function name
            if (it == end || *it != "(")
                throw std::runtime_error("Error: Expected '(' after " + fn);
            ++it;  // skip '('
            double arg = evaluateExpression(it, end);
            nums.push(applyFunction(fn, arg));
            continue;
        }

        // Sub-expression
        if (tok == "(") {
            ++it;
            nums.push(evaluateExpression(it, end));
            continue;
        }

        // Number literal
        if (std::isdigit(tok[0]) || (tok[0] == '-' && tok.size() > 1)) {
            nums.push(std::stod(tok));
            ++it;
            continue;
        }

        // Operator
        if (tok.size() == 1 && precedence.count(tok[0])) {
            char op = tok[0];
            while (!ops.empty()) {
                char top = ops.top();
                int p1 = precedence[op];
                int p2 = precedence[top];
                if ((op == '^' && p1 < p2) ||
                    (op != '^' && p1 <= p2)) {
                    ops.pop();
                    double b = nums.top(); nums.pop();
                    double a = nums.top(); nums.pop();
                    nums.push(applyOperation(a, b, top));
                }
                else break;
            }
            ops.push(op);
            ++it;
            continue;
        }

        throw std::runtime_error("Error: Invalid token '" + tok + "'");
    }

    // Consume ')'
    if (it != end && *it == ")") ++it;

    // Flush remaining ops
    while (!ops.empty()) {
        char op = ops.top(); ops.pop();
        double b = nums.top(); nums.pop();
        double a = nums.top(); nums.pop();
        nums.push(applyOperation(a, b, op));
    }

    if (nums.empty())
        throw std::runtime_error("Error: Empty expression!");
    return nums.top();
}

// Wrapper to evaluate full token list
double evaluate(const std::vector<std::string>& tokens) {
    auto it = tokens.begin();
    return evaluateExpression(it, tokens.end());
}

void interactiveMode(History& history){
    for(;;) {
        std::string line;
        std::getline(std::cin, line);
        // store expression in history
        history.write(line);

        try {
            if (!checkParentheses(line))
                throw std::runtime_error("Error: Mismatched parentheses.");

            auto tokens = tokenize(line);
            double result = evaluate(tokens);
            std::cout << result << "\n";
        }
        catch (const std::exception& e) {
            std::cerr << e.what() << "\n";
        }
    }
}

void argumentMode(History& history, const std::string& arg){
    // store expression in history
    history.write(arg);
    try {
        if (!checkParentheses(arg))
            throw std::runtime_error("Error: Mismatched parentheses.");

        auto tokens = tokenize(arg);
        double result = evaluate(tokens);
        std::cout << result << "\n";
    }
    catch (const std::exception& e) {
        std::cerr << e.what() << "\n";
    }
}

int main(int argc, char* argv[]) {
    History history;
    g_history = &history;

    // Register signal handlers to flush history on interrupt
    std::signal(SIGINT, signalHandler);
    std::signal(SIGTERM, signalHandler);

    // Default to inline-single mode when no flags are provided
    Mode mode = Mode::MODE_INTERACTIVE;

    // If there are arguments but no known modifier flags, treat all argv
    // elements (1..argc-1) as a single expression and evaluate it.
    std::ostringstream oss;
    if (argc > 1) {
        mode = Mode::MODE_ARGUMENT;
        for (int i = 1; i < argc; ++i) {
            if (i > 1) oss << ' ';
            oss << argv[i];
        }
    }

    // Dispatch based on selected mode
    switch (mode) {
    case Mode::MODE_INTERACTIVE:
        interactiveMode(history);
        break;
    case Mode::MODE_ARGUMENT:
        argumentMode(history, oss.str());
        break;
    }

    return 0;
}
